wallet: Identify transactions spending 0-value outputs, and add tests for anchor outputs in a wallet

achow101 opened the pull request:
One of the ways that the wallet would determine if a transaction was sent from the wallet was by checking if the total amount being spent by a transaction from outputs known to the wallet was greater than 0. This has worked fine until recently since there was no reason for 0-value outputs to be created. However, with ephemeral dust and P2A, it is possible to create standard 0-value outputs, and the wallet was not correctly identifying the spends of such outputs. This PR updates `IsFromMe` to only check whether the wallet knows any of the inputs, rather than checking the debit amount of a transaction.

Additionally, a new functional test is added to test for this case, as well as a few other anchor output related scenarios. This also revealed a bug in `sendall` which would cause an assertion error when trying to spend all of the outputs in a wallet that has anchor outputs.

Fixes #33265

achow101 committed:
wallet: Add m_cached_from_me to cache "from me" status

m_cached_from_me is used to track whether a transaction is "from me", i.e. has
any inputs which belong to the wallet. This is held in memory only in
the same way that a transaction's balances are.

achow101 committed:
wallet: Determine IsFromMe by checking for TXOs of inputs

Instead of checking whether the total amount of inputs known by the
wallet is greater than 0, we should be checking for whether the input is
known by the wallet. This enables us to determine whether a transaction
spends an of output with an amount of 0, which is necessary for marking
0-value dust outputs as spent.

achow101 committed:
test: Add a test for anchor outputs in the wallet

achow101 committed:
wallet: Throw an error in sendall if the tx size cannot be calculated

jsarenik commented:
After it gets merged into `master` branch, I would recommend backporting it to current stable release branch which already contains the ephemeral outputs support ([29.x](https://github.com/bitcoin/bitcoin/blob/7cc9a087069bfcdb79a08ce77eb3a60adf9483af/doc/release-notes/release-notes-29.0.md?plain=1#L64}).

instagibbs commented on test/functional/wallet_anchor.py:
```diff
@@ -0,0 +1,133 @@
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+
+    def test_anchor_listunspent(self):
```
this subtest is covering 0-value outputs, not anchors per se. I'd rather keep the tests logically separated to not confused the concepts more than they already are :)

achow101 replied:
Clarified the test name

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -1634,7 +1634,11 @@ bool CWallet::IsMine(const COutPoint& outpoint) const
 bool CWallet::IsFromMe(const CTransaction& tx) const
 {
-    return (GetDebit(tx) > 0);
+    LOCK(cs_wallet);
```
Updated the commit message.

jsarenik commented:
Tested ACK de98ffb

furszy reviewed (commented):
q: is `IsFromMe()` method still relevant? It seems we could directly call `IsMine()` on the inputs, which is also cached now and should be slightly safer for migration.

furszy commented:
> > is `IsFromMe()` method still relevant?
> 
> Yes, transactions that we make are still treated specially.

Where?. `IsFromMe()` seems to be only ever called in `AddToWalletIfInvolvingMe()` and `ApplyMigrationData()` next to the `IsMine()` call. In both places we effectively treat it as `is_mine = IsMine(outputs) || IsFromMe(inputs)`.

achow101 commented:
> Where?

Hmm, I forgot that `CachedTxIsFromMe` does not use `IsFromMe`. Might need to change that in this PR as well.

achow101 committed:
test: Test wallet 'from me' status change

If something is imported into the wallet, it can change the 'from me'
status of a transaction. This status is only visible through
gettransaction's "fee" field which is only shown for transactions that
are 'from me'.

w0xlt reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/33268/commits/69eeb11bc585e92a3b56811c3251f13dba05aa5f with non-blocking nits above.

kannapoix reviewed (commented):
ACK 69eeb11bc585e92a3b56811c3251f13dba05aa5f

I have run the all functional tests and reviewed the code.
I left a nit comment below.

rkrux reviewed (commented):
Concept ACK 69eeb11bc585e92a3b56811c3251f13dba05aa5f

Agree with the intent. Nit in PR description because specifically `IsFromMe` is updated in the PR:

```diff
- One of the ways that the wallet would determine if a transaction belonged to the wallet was by checking
+ One of the ways that the wallet would determine if a transaction was sent from the wallet was by checking
```

achow101 commented on test/functional/wallet_anchor.py:
```diff
@@ -0,0 +1,133 @@
+        assert_equal(utxos[0]["txid"], anchor_txid)
+        assert_equal(utxos[0]["address"], ANCHOR_ADDRESS)
+        assert_equal(utxos[0]["amount"], 0)
+        wallet.gettransaction(anchor_txid)
```
There's nothing to assert, it either error and throws, or it succeeds and returns. The return value can never be false-y.

rkrux replied:
Fair point, this suggestion was not correct.

The intent of the suggestion was more for the cosmetic angle so that the reader doesn't wonder why a standalone RPC is here without the presence of an assertion or usage of the returned value. Maybe there is an opportunity to introduce an `assert_no_rpc_error` helper here but need not be done now.

achow101 commented on src/wallet/receive.cpp:
The test coverage for this commit is provided in the first commit of the PR. The point of that test is to ensure that behavior doesn't change for commit.

rkrux replied:
I see, I had missed the connection between the two commits.

achow101 commented on test/functional/wallet_listtransactions.py:
```diff
@@ -311,6 +317,45 @@ def test_op_return(self):
+            assert_equal(send_res["complete"], True)
+            txid = send_res["txid"]
+            self.nodes[0].syncwithvalidationinterfacequeue()
+            assert "fee" not in wallet.gettransaction(txid)
```
Done as suggested

achow101 commented on test/functional/wallet_anchor.py:
```diff
@@ -0,0 +1,133 @@
+)
+from test_framework.wallet import MiniWallet
+
+ANCHOR_ADDRESS = "bcrt1pfeesnyr2tx"
```
Moved

achow101 commented on test/functional/wallet_anchor.py:
```diff
@@ -0,0 +1,133 @@
+
+        assert_raises_rpc_error(-4, "Missing solving data for estimating transaction size", wallet.send, [{self.default_wallet.getnewaddress(): 0.9999}])
+        assert_raises_rpc_error(-4, "Error: Private keys are disabled for this wallet", wallet.sendtoaddress, self.default_wallet.getnewaddress(), 0.9999)
+        assert_raises_rpc_error(-4, "Unable to determine the size of the transaction, the wallet contains unsolvable descriptors", wallet.sendall, recipients=[self.default_wallet.getnewaddress()], inputs=utxos)
```
Done

achow101 commented on src/wallet/receive.cpp:
```diff
@@ -195,7 +195,9 @@ void CachedTxGetAmounts(const CWallet& wallet, const CWalletTx& wtx,
-    return (CachedTxGetDebit(wallet, wtx, /*avoid_reuse=*/false) > 0);
+    if (wtx.m_cached_from_me.has_value()) return wtx.m_cached_from_me.value();
+    wtx.m_cached_from_me = wallet.IsFromMe(*wtx.tx);
+    return wtx.m_cached_from_me.value();
```
Done

rkrux commented on test/functional/wallet_listtransactions.py:
```diff
@@ -311,6 +317,45 @@ def test_op_return(self):
+
+            if confirm:
+                self.generate(self.nodes[0], 1, sync_fun=self.no_op)
+                # Mock time forward and generate blocks so that the import does not rescan the transaction
+                self.nodes[0].setmocktime(int(time.time()) + MAX_FUTURE_BLOCK_TIME + 1)
+                self.generate(self.nodes[0], 10, sync_fun=self.no_op)
```
I don't believe rescanning is skipped here at the moment. The following assertion passes, whereas putting it in the `unexpected` section throws.

```diff
diff --git a/test/functional/wallet_listtransactions.py b/test/functional/wallet_listtransactions.py
index 714e659465..b590ed95d4 100755
--- a/test/functional/wallet_listtransactions.py
+++ b/test/functional/wallet_listtransactions.py
@@ -352,7 +352,8 @@ class ListTransactionsTest(BitcoinTestFramework):
                 self.nodes[0].setmocktime(int(time.time()) + MAX_FUTURE_BLOCK_TIME + 1)
                 self.generate(self.nodes[0], 10, sync_fun=self.no_op)
 
-            import_res = wallet.importdescriptors([{"desc": descriptor, "timestamp": "now"}])
+            with self.nodes[0].assert_debug_log(expected_msgs=["Rescanning last"]):
+                import_res = wallet.importdescriptors([{"desc": descriptor, "timestamp": "now"}])
             assert_equal(import_res[0]["success"], True)
             # TODO: We should check that the fee matches, but since the transaction spends inputs
             # not known to the wallet, it is incorrectly calculating the fee.
```

rkrux reviewed (approved):
lgtm ACK 113a4228229baedda2a730e097f2d59ad58a4b0d

Generally, I find it easy to miss these `Cached*` functions because they lie in this `wallet/receive.cpp` file that I don't find intuitive; `wallet/transaction.cpp` could be more intuitive because most of these functions update the `CWallet` object somehow - this PR need not do anything about it though.

Ok with addition of `m_cached_from_me` within `WalletTx` in this PR but I find `Cached*` functions are too many to keep track of.

enirox001 reviewed (commented):
Tested ACK 113a422. Ran the full functional test suite including `wallet_anchor.py`; all tests passed. Fix for 0 value anchor detection and sendall size errors looks good. LGTM.

furszy reviewed (commented):
ACK 113a4228229baedda2a730e097f2d59ad58a4b0d

fanquake commented:
Backported to 29.x in #34268.
